File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionCsv.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.*;
import co.nlighten.jsontransform.formats.csv.CsvFormat;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/*
 * @see CsvFormat for arguments explanation
 * For tests
 * @see TransformerFunctionCsvTest
 */
public class TransformerFunctionCsv extends TransformerFunction {
    static final Logger logger = LoggerFactory.getLogger(TransformerFunctionCsv.class);

    public TransformerFunctionCsv() {
        super(FunctionDescription.of(
            Map.of(
            "no_headers", ArgumentType.of(ArgType.Boolean).position(0).defaultBoolean(false),
            "force_quote", ArgumentType.of(ArgType.Boolean).position(1).defaultBoolean(false),
            "separator", ArgumentType.of(ArgType.String).position(2).defaultString(","),
            "names", ArgumentType.of(ArgType.Array).position(3).defaultIsNull(true)
            )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        var streamer = context.getJsonElementStreamer(null);
        try {
            if (streamer == null)
                return null;
            var names = context.getJsonArray("names");
            var noHeaders = context.getBoolean("no_headers");
            var forceQuote = context.getBoolean("force_quote");
            var separator = context.getString("separator");
            List<String> namesList = null;
            var adapter = context.getAdapter();
            if (names != null) {
                var stringStream = (Stream<String>) adapter.stream(names).map(context::getAsString);
                namesList = stringStream.collect(Collectors.toList());
            }
            return new CsvFormat(
                    adapter,
                    namesList,
                    noHeaders,
                    forceQuote,
                    separator)
                    .serialize(streamer.toJsonArray());
        } catch (Exception e) {
            logger.warn(context.getAlias() + " failed", e);
            return null;
        }
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionCsv.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import { getAsString } from "../JsonHelpers";
import CsvFormat from "../formats/csv/CsvFormat";

class TransformerFunctionCsv extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        no_headers: { type: ArgType.Boolean, position: 0, defaultBoolean: false },
        force_quote: { type: ArgType.Boolean, position: 1, defaultBoolean: false },
        separator: { type: ArgType.String, position: 2, defaultString: "," },
        names: { type: ArgType.Array, position: 3, defaultIsNull: true },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const streamer = await context.getJsonElementStreamer(null);
    try {
      if (streamer == null) return null;
      const names = await context.getJsonArray("names");
      const noHeaders = await context.getBoolean("no_headers");
      const forceQuote = await context.getBoolean("force_quote");
      const separator = await context.getString("separator");
      const namesList = names?.map(el => getAsString(el) ?? "");
      return new CsvFormat(namesList, noHeaders, forceQuote, separator).serialize(await streamer.toJsonArray());
    } catch (e: any) {
      console.warn(context.getAlias() + " failed", e);
      return null;
    }
  }
}

export default TransformerFunctionCsv;
